### PR TITLE
Fix tooltips

### LINKED ISSUE #17023: Tooltips for messages with an explicit loc broken
The changes made in #16978 don't display tooltips correctly when the warning has a `loc` value.  (I'm not sure that PR was intended to fix this case.)

I added a command that generates a warning with a `loc`.  ~The first time you run it,~ If you run it with 2 `Forward`s the tooltip doesn't show up correctly:

image

~Resetting Coq and rerunning the script in the same buffer~, If you ~undo and rerun `Foo`~ run both the `Goal` and `Foo` with a single `Run to End`, the tooltip doesn't appear at all.  (You also get this with 2 `Forward`s if `Foo` is not the last line in the buffer.):

image

My test command:

```
      | IDENT "Foo"; x = string -> {
                                        Printf.eprintf "Some bp ep %d %d \n%!" loc.Loc.bp loc.ep;
                                        let f = CWarnings.create ~name:"Foo" ~category:"scope"
                                            (fun () ->
                                             Pp.strbrk "Foo message") in
                                             f ~loc ();
                                             VernacAbort }
```

Attn: @ejgallego

### PR DESCRIPTION
**Pull request by jfehrle**
I believe this fixes all the issues in #17023.

First commit:
- fixes the broken code in [#16978](https://github.com/coq/coq/issues/16978) that assumed Coq sends `lineno` as part of locations, which it does not.  That meant that no tooltips would be appear at all, e.g. in `Pff.v`.  This part should go into 8.17.
- handles the case where the position of a sentence changes after it's been sent to the server (described [here](https://github.com/coq/coq/issues/17023#issuecomment-1367510209))

Second commit:
A workaround to handle warning and error tooltips on commands correctly without changing the protocol.  This is likely a very old bug.  Optional for 8.17, I can split it out if desired.  The issues are:
- The warning message is sent before the `Good` message for `add`, which sets the sentence id.  Previously the message was treated as a message on an unknown sentence and thus not displayed as a tooltip.
- In other cases, it appears the message is sent with the wrong message id.  For this, there is a check whether the range in the Loc is within the range of the sentence.  If not, the tooltip is attached to the sentence passed in the last `add`, which appears to work correctly.

BTW, I will be travelling Jan 3-10 and thus less available.

Fixes: #17023

Attn @ejgallego

### CODE REVIEW

**Comment by ejgallego** [comment]
Definitively this should not be the approach implemented in master, instead, the protocol should be fixed to send non-bogus locations. It makes little sense to add a quite contrived ad-hoc cache client-side to workaround what amounts to a protocol deficiency.

For 8.17 indeed we could just implement a more conservative fix I think.

File: ide/coqide/coqOps.ml [hunk 1]
```diff
@@ -34,9 +34,11 @@ end
 
 module SentenceId : sig
 
-  type sentence = private {
```
Comment (ejgallego) [inline ide/coqide/coqOps.ml:37]: Don't remove the private here.

File: ide/coqide/coqOps.ml [hunk 2]
```diff
@@ -34,9 +34,11 @@ end
...
+  type sentence = {
     start : GText.mark;
     stop : GText.mark;
+    mutable bp : int;
+    mutable ep : int;
```
Comment (ejgallego) [inline ide/coqide/coqOps.ml:41]: Please document what this is for.

File: ide/coqide/coqOps.ml [hunk 3]
```diff
@@ -138,11 +144,27 @@ end = struct
...
+
 (* Given a Coq loc, convert it to a pair of iterators start / end in
    the buffer. *)
-let coq_loc_to_gtk_offset (buffer : GText.buffer) loc =
-  buffer#get_iter_at_byte ~line:loc.Loc.line_nb (loc.bp - loc.bol_pos),
-  buffer#get_iter_at_byte ~line:loc.Loc.line_nb_last (loc.ep - loc.bol_pos_last)
+let coq_loc_to_gtk_offset (buffer : GText.buffer) loc sentence =
+  (* n_bytes is invariant even if the sentence was moved though the
+     buffer offset will differ *)
+  let adjust byte_off =
```
Comment (ejgallego) [inline ide/coqide/coqOps.ml:164]: The whole approach here seems bugged to me, the GTK editor component already keeps track of relocation right? If so, we should not have to do this process ourselves.

File: ide/coqide/coqOps.ml [hunk 4]
```diff
@@ -567,6 +589,15 @@ object(self)
...
         try Some (Doc.find_map document finder)
         with Not_found -> None in
+      (* When a command generates a warning or error, the warning/error
+         message is returned before the Processed message that provides the
+         sentence id.  There can only be a single sentence in the document
+         with an id of None, which must be at the tip.  A protocol design
+         weakness.  Not an issue for tactics. *)
+      let sentence = match sentence with
+        | None -> (try Some (None, Doc.tip_data document) with Doc.Empty -> None)
+        | Some (sid, sent) -> sentence
```
Comment (ejgallego) [inline ide/coqide/coqOps.ml:599]: This seems nice! Should go in its own PR IMHO tho.

File: ide/coqide/coqOps.ml [hunk 5]
```diff
@@ -578,6 +609,19 @@ object(self)
           Minilib.log ("Unsupported feedback message " ^ mtype ^ " with a sentence")
       in
       let log ?id s = log_pp ?id (Pp.str s) in
+      (* When a command generates a warning or error, the message sent has an
+         incorrect sentence id.  If bp..ep is out of range for the specified
+         sentence, assume is it is from the last-sent "add" request *)
```
Comment (ejgallego) [inline ide/coqide/coqOps.ml:614]: How come the sentence id is wrong, you mean it lack the Id?

File: ide/coqide/coqOps.ml [hunk 6]
```diff
@@ -578,6 +609,19 @@ object(self)
...
+      let handle_bad_loc loc id sentence =
+        match loc with
+        | Some loc ->
+          if loc.Loc.bp < sentence.bp || loc.Loc.ep > sentence.ep then begin
+              log_pp ?id Pp.(str "WarningMsg/ErrorMsg with bad loc, using tip instead");
```
Comment (ejgallego) [inline ide/coqide/coqOps.ml:619]: This should never happen in the first place, so instead of having this ugly workaround, make sure it doesn't ever happen.

**Review by ejgallego — CHANGES_REQUESTED** [review]

File: ide/coqide/coqOps.ml (hunk 4 above)
Comment (jfehrle) [reply ide/coqide/coqOps.ml:599]: As I say in the PR description, "[the second commit is] optional for 8.17, I can split it out if desired."

File: ide/coqide/coqOps.ml (hunk 5 above)
Comment (jfehrle) [reply ide/coqide/coqOps.ml:614]: AFAIK every feedback message includes a sentence id (`state_id` below), however if CoqIDE doesn't have that state id in its document, it's considered unknown.  I found that both this and the code at 592-600 was needed for differing use cases.  I would have repeat some testing to describe exactly what was happening.

```
<feedback object="state" route="0">
  <state_id val="2"/>
  <feedback_content val="message">
    <message>
      <message_level val="warning"/>
      <option val="some">
        <loc start="11" stop="16"/>
      </option>
```

File: ide/coqide/coqOps.ml (hunk 6 above)
Comment (jfehrle) [reply ide/coqide/coqOps.ml:619]: I don't know enough about Stm to know where to begin to fix it in there.

File: ide/coqide/coqOps.ml (hunk 3 above)
Comment (jfehrle) [reply ide/coqide/coqOps.ml:164]: As I say in the description, "[This] handles the case where the position of a sentence changes after it's been sent to the server (described [here](https://github.com/coq/coq/issues/17023#issuecomment-1367510209))."

When you send a sentence to the server, the bp/ep values for returned Locs in that sentence are based on (relative to) the original bp that was passed in the `add` call.  If the sentence moves, converting the new GTK position to bp will be different than the bp sent in the `add` command and thus incorrect.  Adding the line number to the protocol won't help--you need to remember the original value.

**Comment by jfehrle** [comment]
> Definitively this should not be the approach implemented in master, instead, the protocol should be fixed to send non-bogus locations.  It makes little sense to add a quite contrived ad-hoc cache client-side to workaround what amounts to a protocol deficiency.

I have no objection to adding the line number in the transmitted `loc`, but I'd do that a bit later.  However, as I note above, it does not address one case.

> For 8.17 indeed we could just implement a more conservative fix I think.

Go ahead, make my day.  I've tested this a bunch and I believe it gives the correct behavior for all the cases I describe.  Perhaps you or one of the others will kick the tires and let me know if there are any remaining problems.  If anyone finds a bug in it, I will try to fix it.

If this indeed gives the correct behavior, we should merge at least the first commit into 8.17 since #16978 is rather broken (described [here](https://github.com/coq/coq/issues/17023#issuecomment-1368513684)).

File: ide/coqide/coqOps.ml (hunk 3 above)
Comment (ejgallego) [reply ide/coqide/coqOps.ml:164]: What's the invariant that must hold regarding `bp` and `start` in a sentence?

Comment (jfehrle) [reply ide/coqide/coqOps.ml:164]: Suppose you send a sentence to the server in an "add" request with sentence.bp = 100 and sentence.ep = 200.  The returned loc (say bp=110, ep=120) for a warning will be a subrange of 100..200.  These offsets won't change even if the sentence moves in the buffer.

If the user later enters or removes characters before the sentence and a similar warning is received for the sentence, the 10 and 20 byte offsets relative to sentence.start are still correct.  All that's needed is to figure out how many GTK chars correspond to the first 10 or 20 bytes offset of the sentence, i.e. by examining the first 10 or 20 bytes of chars after `start`.

Maybe "invariant" is not the ideal word.

I think it's pretty straightforward once you grasp it.  More words may just confuse. Probably better explained with a picture but not so easy to create one.  Hope this helps, let me know.  _Tuum passibus variari potest._

**Comment by Zimmi48** [comment]
The issue was fixed by other PRs.
